File: src/test/java/aww/yeah/TwoHundredControllerTest.java
package aww.yeah;


import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.core.io.InputStreamResource;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;

import java.util.Random;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
public class TwoHundredControllerTest {

    private TwoHundredController twoHundredController;

    @Mock
    private Random random;

    @BeforeEach
    public void setUp() throws Exception {
        twoHundredController = new TwoHundredController(random);
    }

    @Test
    void respondToAnything_shouldReturnJson_whenRandomIsNotZero() throws Exception {
        when(random.nextInt(anyInt())).thenReturn(1);
        ResponseEntity responseEntity = twoHundredController.respondToAnything();

        assertThat(responseEntity.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(responseEntity.getBody()).isEqualTo("{\"success\":\"Aww Yeah\"}");
    }

    @Test
    void respondToAnything_shouldReturnGif_whenRandomIsZero() throws Exception {
        when(random.nextInt(anyInt())).thenReturn(0);
        ResponseEntity responseEntity = twoHundredController.respondToAnything();

        assertThat(responseEntity.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(((InputStreamResource)responseEntity.getBody()).getInputStream())
                .hasSameContentAs(new InputStreamResource(ClassLoader.getSystemResourceAsStream("aww_yeah.gif")).getInputStream());
    }

    @Test
    void getErrorPath_shouldReturn_slashError() throws Exception {
        assertThat(twoHundredController.getErrorPath()).isEqualTo("/error");
    }


}

File: src/test/kotlin/aww/yeah/TwoHundredControllerTest.kt
package aww.yeah

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.core.io.InputStreamResource
import org.springframework.core.io.InputStreamSource
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType

internal class TwoHundredControllerTest {

    @Test
    internal fun `respondToAnything() should return json when show gif is false`() {

        val twoHundredController = TwoHundredController { false }

        val responseEntity = twoHundredController.respondToAnything()

        assertThat(responseEntity.statusCode).isEqualTo(HttpStatus.OK)
        assertThat(responseEntity.headers.contentType).isEqualTo(MediaType.APPLICATION_JSON)
        assertThat(responseEntity.body).isEqualTo(Success())
    }

    @Test
    internal fun `respondToAnything() should return gif when show gif is true`() {

        val twoHundredController = TwoHundredController { true }

        val responseEntity = twoHundredController.respondToAnything()

        assertThat(responseEntity.statusCode).isEqualTo(HttpStatus.OK)
        assertThat(responseEntity.headers.contentType).isEqualTo(MediaType.IMAGE_GIF)
        assertThat((responseEntity.body as InputStreamSource).inputStream)
                .hasSameContentAs(
                        InputStreamResource(
                                ClassLoader.getSystemResourceAsStream("aww_yeah.gif")!!)
                                .inputStream)
    }
}

